File: nats/python/receiver_async_test.py
#!/usr/bin/env python3
"""NATS Python Receiver - Async"""
import sys
import signal
import nats
import asyncio
from pathlib import Path

# Add utils to path
repo_root = Path(__file__).parent.parent.parent
sys.path.insert(0, str(repo_root / 'utils' / 'python'))

from message_helpers import *

running = True

def signal_handler(sig, frame):
    global running
    running = False


async def message_handler(msg, receiver_id):
    """Handle incoming message asynchronously."""
    request_str = msg.data
    
    # Parse message
    request_envelope = parse_envelope(request_str)
    message_id = request_envelope.message_id
    print(f" [x] [ASYNC] Received message {message_id}")
    
    # Create ACK
    response = create_ack_from_envelope(request_envelope, str(receiver_id))
    setattr(response, 'async', True)
    resp_str = serialize_envelope(response)
    
    # Send reply
    await msg.respond(resp_str)


async def run(receiver_id):
    nc = await nats.connect("nats://localhost:4222")
    
    subject = f"test.subject.{receiver_id}"
    
    async def cb(msg):
        await message_handler(msg, receiver_id)

    # Subscribe with handler
    await nc.subscribe(subject, cb=cb)
    
    print(f" [*] [ASYNC] Receiver {receiver_id} subscribed to {subject}")
    
    # Keep running
    while running:
        await asyncio.sleep(0.1)
    
    print(f" [x] [ASYNC] Receiver {receiver_id} shutting down")
    await nc.close()


def main():
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('--id', type=int, default=0)
    args = parser.parse_args()
    
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)
    
    asyncio.run(run(args.id))


if __name__ == "__main__":
    main()

File: nats/cpp/receiver_async_test.cpp
#include <nats/nats.h>
#include <iostream>
#include <string>
#include <signal.h>
#include <thread>
#include <atomic>
#include "../../utils/cpp/message_helpers.hpp"

using messaging::MessageEnvelope;
using message_helpers::get_current_time_ms;

std::atomic<bool> running(true);

void signal_handler(int sig) {
    running = false;
}

void onMsg(natsConnection *nc, natsSubscription *sub, natsMsg *msg, void *closure) {
    int *receiver_id = (int*)closure;

    std::string request_str(natsMsg_GetData(msg), natsMsg_GetDataLength(msg));
    
    MessageEnvelope request_envelope;
    if (message_helpers::parse_envelope(request_str, request_envelope)) {
        std::string message_id = request_envelope.message_id();
        std::cout << " [x] [ASYNC] Received message " << message_id << std::endl;

        // Create ACK
        MessageEnvelope response = message_helpers::create_ack_from_envelope(
            request_envelope,
            std::to_string(*receiver_id)
        );
        response.set_async(true);
        std::string resp_str = message_helpers::serialize_envelope(response);
        
        // Send reply
        if (natsMsg_GetReply(msg)) {
            natsConnection_Publish(nc, natsMsg_GetReply(msg), resp_str.c_str(), (int)resp_str.size());
        }
    }

    natsMsg_Destroy(msg);
}

int main(int argc, char* argv[]) {
    int receiver_id = 0;
    
    for (int i = 1; i < argc; ++i) {
        if (std::string(argv[i]) == "--id" && i + 1 < argc) {
            receiver_id = std::stoi(argv[i+1]);
        }
    }

    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);

    natsConnection *conn = NULL;
    natsStatus s = natsConnection_ConnectTo(&conn, "nats://localhost:4222");
    if (s != NATS_OK) {
        std::cerr << "Failed to connect: " << natsStatus_GetText(s) << std::endl;
        return 1;
    }

    std::string subject = "test.subject." + std::to_string(receiver_id);
    natsSubscription *sub = NULL;
    s = natsConnection_Subscribe(&sub, conn, subject.c_str(), onMsg, &receiver_id);
    if (s != NATS_OK) {
        std::cerr << "Failed to subscribe: " << natsStatus_GetText(s) << std::endl;
        natsConnection_Destroy(conn);
        return 1;
    }

    std::cout << " [*] [ASYNC] Receiver " << receiver_id << " subscribed to " << subject << std::endl;

    while (running) {
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    std::cout << " [x] [ASYNC] Receiver " << receiver_id << " shutting down" << std::endl;

    natsSubscription_Destroy(sub);
    natsConnection_Destroy(conn);
    return 0;
}
